pkg/compactor/split_merge_grouper.go

```diff
@@ -177,13 +177,25 @@ func planCompaction(userID string, blocks []*block.Meta, ranges []int64, shardCo
 
 	// Jobs will be sorted later using configured job sorting algorithm.
 	// Here we sort them by sharding key, to keep the output stable for testing.
-	sort.SliceStable(jobs, func(i, j int) bool {
-		if iKey, jKey := jobs[i].shardingKey(), jobs[j].shardingKey(); iKey != jKey {
-			return iKey < jKey
+	slices.SortStableFunc(jobs, func(a, b *job) int {
+		aKey, bKey := a.shardingKey(), b.shardingKey()
+		if aKey != bKey {
+			if aKey < bKey {
+				return -1
+			}
+			return 1
 		}
 
 		// The sharding key could be equal but external labels can still be different.
-		return defaultGroupKeyWithoutShardID(jobs[i].blocks[0].Thanos) < defaultGroupKeyWithoutShardID(jobs[j].blocks[0].Thanos)
+		aGroupKey := defaultGroupKeyWithoutShardID(a.blocks[0].Thanos)
+		bGroupKey := defaultGroupKeyWithoutShardID(b.blocks[0].Thanos)
+		if aGroupKey < bGroupKey {
+			return -1
+		}
+		if aGroupKey > bGroupKey {
+			return 1
+		}
+		return 0
```

Comment:
(I haven't checked, but I assume this is a string)
```suggestion
		return strings.Compare(aGroupKey, bGroupKey)
```